test/action_test.dart

```diff
@@ -199,18 +197,233 @@ void main() {
+        await Future(() => {});
+        expect(listenerInvoked, isTrue);
+        expect(listenerCompleted, isFalse);
+
+        // Completed in next next scheduled event.
+        await Future(() => {});
+        expect(listenerCompleted, isTrue);
+      });
+
+      test('should complete future after listeners complete', () async {
+        var asyncListenerCompleted = false;
         action.listen((_) async {
-          asyncCompleter.complete();
+          await Future.delayed(const Duration(milliseconds: 100), () {
+            asyncListenerCompleted = true;
+          });
         });
+
+        Future<dynamic>? future = action('payload');
+        expect(asyncListenerCompleted, isFalse);
+
+        await future;
+        expect(asyncListenerCompleted, isTrue);
+      });
+
+      test('should surface errors in listeners', () {
+        var actionWithError = action..listen((_) => throw UnimplementedError());
+        expect(actionWithError('payload'), throwsUnimplementedError);
+      });
+    });
+
+    group('listen', () {
+      test('should stop listening when subscription is canceled', () async {
+        var listened = false;
+        var subscription = action.listen((_) => listened = true);
+
+        await action('payload');
+        expect(listened, isTrue);
+
+        listened = false;
+        subscription.cancel();
+        await action('payload');
+        expect(listened, isFalse);
+      });
+
+      test('should stop listening when listeners are cleared', () async {
+        var listened = false;
+        action.listen((_) => listened = true);
+
+        await action('payload');
+        expect(listened, isTrue);
+
+        listened = false;
+        await action.dispose();
+        await action('payload');
+        expect(listened, isFalse);
+      });
+
+      test('should stop listening when actions are disposed', () async {
+        var listened = false;
+        action.listen((_) => listened = true);
+
+        await action('payload');
+        expect(listened, isTrue);
+
+        listened = false;
+        await action.dispose();
+        await action('payload');
+        expect(listened, isFalse);
+      });
+    });
+
+    group('Null typed', () {
+      late Action2<Null> nullAction;
+
+      setUp(() {
+        nullAction = Action2<Null>();
+      });
+
+      test('should support dispatch with a null payload', () async {
+        Completer c = Completer();
+        nullAction.listen((payload) {
+          expect(payload, isNull);
+          c.complete();
+        });
+
+        await nullAction(null);
+        return c.future;
+      });
+    });
+
+    group('void typed', () {
+      var voidAction = Action2<void>();
+
+      test('should support dispatch with a null payload', () async {
+        Completer c = Completer();
+        voidAction.listen((_) {
+          c.complete();
+        });
+
+        await voidAction(null);
+        return c.future;
+      });
+    });
+
+    group('benchmarks', () {
```

Comment:
This test was copied from the original Action class and is skipped. I'm not sure why. When un-skipped it passes.

Should we un-skip?